src/utils/filter-utils.js

```diff
@@ -369,6 +435,74 @@ export function filterData(dataset, filters) {
   return {data: filtered, filteredIndex, filteredIndexForDomain};
 }
 
+/**
+ * This returns the points to evaluate in polygon filter
+ * @param layer
+ * @param data
+ * @return {*[][]}
+ */
+export const getLayerPointsToFilter = (layer, data) => {
+  switch (layer.type) {
+    case LAYER_TYPES.arc:
+    case LAYER_TYPES.line:
+      const {
+        lng0,
+        lat0,
+        lng1,
+        lat1
+      } = layer.config.columns;
+      return [
+        [data[lng0.fieldIdx], data[lat0.fieldIdx]],
+        [data[lng1.fieldIdx], data[lat1.fieldIdx]]
+      ];
+    default:
+      const {lng, lat} = layer.config.columns;
+      return [
+        [data[lng.fieldIdx], data[lat.fieldIdx]]
+      ];
+  }
+};
+
+const filterDataMatchers = {
+  [FILTER_TYPES.range]: (data, filter, index, field) => {
+    const val = field ? data[field.tableFieldIndex - 1] : null;
+    return isInRange(val, filter.value)
+  },
+  [FILTER_TYPES.timeRange]: (data, filter, index, field) => {
+    const val = field ? data[field.tableFieldIndex - 1] : null;
+    const timeVal = field && field.filterProp && Array.isArray(field.filterProp.mappedValue)
+      ? field.filterProp.mappedValue[index]
+      : moment.utc(val).valueOf();
+
+    return isInRange(timeVal, filter.value);
+  },
+  [FILTER_TYPES.multiSelect]: (data, filter, index, field) => {
+    const val = field ? data[field.tableFieldIndex - 1] : null;
+    return filter.value.includes(val)
+  },
+  [FILTER_TYPES.select]: (data, filter, index, field) => {
+    const val = field ? data[field.tableFieldIndex - 1] : null;
+    return val === filter.value
+  },
+  // layers contain only layers for the current dataset
+  [FILTER_TYPES.polygon]: (data, filter, index, field, layers) => {
+    if (!(layers || layers.length === 0)) {
```

Comment:
This function is called on every data point, `currentLayers` should be passed into this function. Instead of creating it on every pass.